### PR TITLE
refactors faxes a little

### PR DESCRIPTION
**Pull request by harryob**
get it? refaxtored? like the branch name?

:cl:
refactor: cleaned up duplicated fax code
/:cl:

### CODE REVIEW

File: code/game/machinery/fax_machine.dm [hunk 1]
```diff
@@ -1,9 +1,319 @@
...
+	for(var/client/C in GLOB.admins)
+		if((R_ADMIN|R_MOD) & C.admin_holder.rights)
+			if(msg_admin)
+				to_chat(C, msg_admin)
+			else
+				to_chat(C, msg_ghost)
+			SEND_SOUND(C, 'sound/effects/sos-morse-code.ogg')
+	if(msg_ghost)
+		for(var/i in GLOB.observer_list)
+			var/mob/dead/observer/g = i
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:89]: you can typecast the var in the for loop; no one letter vars

File: code/game/machinery/fax_machine.dm [hunk 2]
```diff
@@ -1,9 +1,319 @@
...
+/**
+ * Announces a fax to ghosts and admins.
+ *
+ * Arguments:
+ * msg_admin: required, the message to send to admins
+ * msg_ghost: optional, the message to send to ghosts
+ */
+/datum/communications_network/proc/announce_fax(msg_admin, msg_ghost)
+	log_admin(msg_admin) //Always irked me the replies do show up but the faxes themselves don't
+	for(var/client/C in GLOB.admins)
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:80]: as anything; one letter var

File: code/game/machinery/fax_machine.dm [hunk 3]
```diff
@@ -1,9 +1,319 @@
...
+				to_chat(C, msg_admin)
+			else
+				to_chat(C, msg_ghost)
+			SEND_SOUND(C, 'sound/effects/sos-morse-code.ogg')
+	if(msg_ghost)
+		for(var/i in GLOB.observer_list)
+			var/mob/dead/observer/g = i
+			if(!g.client)
+				continue
+			var/client/C = g.client
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:92]: one letter var

File: code/game/machinery/fax_machine.dm [hunk 4]
```diff
@@ -1,9 +1,319 @@
...
+			if(C && C.admin_holder)
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:93]: C?.admin_holder

File: code/game/machinery/fax_machine.dm [hunk 5]
```diff
@@ -1,9 +1,319 @@
...
+GLOBAL_DATUM_INIT(communications_network, /datum/communications_network, new)
+
+/datum/communications_network
+	/// All uplinks available, uplink name -> uplink
+	var/list/uplinks
+	/// All faxes, network name -> fax
+	var/list/local_networks
+
+/datum/communications_network/New()
+	for(var/datum/uplink/linking as anything in subtypesof(/datum/uplink))
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:20]: a better type name than `/datum/uplink` might be nice, e.g. `/datum/fax_faction` or smth

File: code/game/machinery/fax_machine.dm [hunk 6]
```diff
@@ -1,9 +1,319 @@
...
+ * Unregisters a department from the a network.
+ *
+ * Arguments:
+ * network: required, the network to remove the department from
+ * department: required, the department to remove
+ */
+/datum/communications_network/proc/unregister_department(network, department)
+	LAZYREMOVEASSOC(local_networks, network, department)
+
+/datum/communications_network/proc/send_fax(sent, sent_name, mob/sender, target_department, network, obj/structure/machinery/faxmachine/origin, datum/uplink/authoriser)
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:68]: codedoc

File: code/game/machinery/fax_machine.dm [hunk 7]
```diff
@@ -1,9 +1,319 @@
...
+ * Arguments:
+ * user: required, the /mob of the admin attempting to send a fax.
+ * replying_to: optional, the /mob that sent the fax the admin is replying to
+ * origin: optional, the /obj/structure/machinery/faxmachine that sent the fax the admin is replying to
+ * from_uplink: optional, the /datum/uplink that is sending the fax
+ * from_network: optional, the network from /datum/uplink that is sending the fax
+ */
+/datum/communications_network/proc/admin_fax(mob/user, mob/replying_to, obj/structure/machinery/faxmachine/origin, datum/uplink/from_uplink, from_network)
+	if(!from_uplink)
+		var/options = tgui_input_list(user, "Which uplink to appear from?", "Available Uplinks", uplinks)
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:111]: this is why I'd like it named something better; an admin won't know what an "uplink" is at a glance

File: code/game/machinery/fax_machine.dm [hunk 8]
```diff
@@ -1,9 +1,319 @@
...
+		if(!options)
+			return
+		from_uplink = uplinks[options]
+
+	if(!from_network)
+		from_network = tgui_input_list(user, "Which network to appear from?", "Available Networks", from_uplink.networks)
+		if(!from_network)
+			return
+
+	var/list/template_choices = from_uplink.template_choices.Copy()
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:121]: why does this need to be copied?

File: code/game/machinery/fax_machine.dm [hunk 9]
```diff
@@ -1,9 +1,319 @@
...
+			if(!addressed_to)
+				return
+			if(addressed_to == "Sender")
+				addressed_to = replying_to.real_name
+			else
+				addressed_to = tgui_input_text(user, "Who should it be addressed to?")
+		else
+			addressed_to = tgui_input_text(user, "Who should it be sent to?")
+
+		var/message_body = tgui_input_text(user, "Enter message body - HTML is allowed here.", encode = FALSE, multiline = TRUE)
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:149]: warning about no bbcode?

File: code/game/machinery/fax_machine.dm [hunk 10]
```diff
@@ -1,9 +1,319 @@
...
+	if(send_choice != "Send")
+		return
+
+	var/customname = tgui_input_text(user, "Pick a title for the report", "Title")
+	if(!customname)
+		return
+
+	send_fax(fax_message, customname, target_department = origin.department, network = from_network, authoriser = from_uplink)
+
+/datum/received_fax
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:170]: split into its own file

File: code/game/machinery/fax_machine.dm [hunk 11]
```diff
@@ -1,9 +1,319 @@
...
+ * user: required, as /mob
+ */
+/datum/received_fax/proc/reply(mob/user)
+	var/mob/living/carbon/human/replying_to = sender.resolve()
+	var/obj/structure/machinery/faxmachine/origin = origin_fax.resolve()
+	var/datum/uplink/sent_to_uplink = uplink.resolve()
+
+	GLOB.communications_network.admin_fax(user, replying_to, origin, sent_to_uplink, network)
+
+/datum/uplink
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:221]: ditto

File: code/game/machinery/fax_machine.dm [hunk 12]
```diff
@@ -1,9 +1,319 @@
...
+	networks = list(NETWORK_HC, NETWORK_PROVOST)
+	stamp_icon = "paper_stamp-uscm"
+	ghost_color = "#006100"
+	template_choices = list("USCM High Command", "Office of the Provost General")
+
+/datum/uplink/uscm/generate_template(addressed_to, subject, message_body, sent_by, sent_title)
+	return generate_templated_fax(FALSE, "USCM CENTRAL COMMAND", subject, addressed_to, message_body,sent_by, sent_title, "United States Colonial Marine Corps")
 
 //This fax machine will become a colonial one after I have mapped it onto the Almayer.
 /obj/structure/machinery/faxmachine
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:319]: ditto

File: code/game/machinery/fax_machine.dm [hunk 13]
```diff
@@ -102,14 +418,34 @@ var/list/alldepartments = list()
...
+	if(target_department != department && target_department != "General Broadcast")
+		return
+
+	if(inoperable())
+		return
+
+	flick("faxreceive", src)
+	addtimer(CALLBACK(src, PROC_REF(print_paper), sent, sent_name, authoriser), 2 SECONDS)
+
+/obj/structure/machinery/faxmachine/proc/print_paper(sent, sent_name, datum/uplink/authoriser)
```
Comment (Zonespace27) [inline code/game/machinery/fax_machine.dm:434]: codedoc

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/game/machinery/fax_machine.dm (hunk 8 above)
Comment (harryob) [reply code/game/machinery/fax_machine.dm:121]: because i was mutating it and then i stopped mutating it

**Pull request closed without merging**